File: a37주차/택배상자.cpp
#include <string>
#include <vector>
#include <stack>
#include <queue>
#include <iostream>

using namespace std;
queue<int> origin;
stack<int> temp;

bool make_datastack()
{
    if (origin.size())
    {
        temp.push(origin.front());
        origin.pop();
        return true;
    }

    return false;
}

int solution(vector<int> order) {
    int answer = 0;

    for (int i = 1; i <= order.size(); ++i) //1. c를 만족하는 넣어야 하는 순서(1,2,3,,,n)를 가진 큐로 모범 벨트 구현
    {
        origin.push(i);
    }

    for (int i = 0; i < order.size(); ++i)
    {
        if (temp.empty()) //2. 보조 벨트가 비어있으면 모범 벨트의 요소를 하나 넣고, 모범&임시 벨트의 요소 중에 메인 벨트의 요소와 일치하는 지 확인
            if (!make_datastack())
                return answer;


        while (temp.top() != order[i]) // 3. 어느 쪽도 일치 하지 않으면 현재 실어진 택배 개수 반환, 일치하면 이어서 메인 벨트 탐색
            if (!make_datastack())
                return answer;
                   
        temp.pop();
        ++answer;
    }

    return answer;
}

File: a81주차/택배상자.java
import java.util.*;

class Solution {
    
    LinkedList<Integer> origin0;
    Stack<Integer> temp0;
    
    boolean makeDatastack()
    { //4.order의 모든 원소를 맞췄거나, 더이상 메인 벨트에서 원소를 꺼낼 수 없을 때 원소 맞춘 개수 반환
        if (origin0.size() > 0)
        {
            temp0.push(origin0.pollFirst());
            return true;
        }

        return false;
    }
    
    public int solution(int[] order) {
        
        int answer = 0;
        origin0 = new LinkedList<>();
        temp0 = new Stack<>();
        
        int ol = order.length;
        for(int i = 1; i <= ol; ++i)
        {//1.queue(메인 벨트)에 1~n까지의 원소 삽입
            origin0.add(i);
        }
        
        for (int i = 0; i < ol; ++i)
        {
            if (temp0.empty())
            { //2.stack(보조벨트)가 비어있으면 메인벨트에서 원소 하나 빼서 보조벨트에 넣기
                if (!makeDatastack()){return answer;}
            }

            while (temp0.peek() != order[i])
            { //3.보조벨트에서 꺼낸 원소가 order의 원소와 같을 때 까지 반복
                if (!makeDatastack()){return answer;}
            }
                   
            temp0.pop();
            ++answer;
        }

        return answer;
    }
    
}
